Pal 884 palisade readers update documentation (#67)
gchq/Palisade-readers: README.md

This is then used to send a request to the Data Service to retrieve the data.
The Reader library provides the functionality for the Data Service to retrieve the data.
This information is then used to return an output stream of the for the service to process the data before returning it to the client.

In terms of code, there is one interface in the Data Service that is core to this library, the `DataReader`.
This provides the basis for all the solutions used in retrieving the data.

The writer in the Data Service class `SerialisingResponseWriter` and the reader implementation in the abstract class `SerialisedDataReader` provide the implementation for the data retrieval.
The class [HadoopDataReader](hadoop-reader/src/main/java/uk/gov/gchq/palisade/reader/HadoopDataReader.java) is the reader implementation for Hadoop.
In the `SerialisingResponseWriter.write` method, these classes are used to first retrieve the references to the data plus the rules that will be used for filtering.
This is used to set up a processing stream for the client's request.

The `SerialisedDataReader` will retrieve the data as an input stream.

> `SerialisingResponseWriter` will then deserialise the stream into records where the rules are applied to redact or mask the records.

Reviewer: ```suggestion
The `SerialisingResponseWriter` will then deserialise the stream into records where the rules are applied to redact or mask the records.

```

This information is then used to return an output stream of the filtered data to the client.

In terms of code, there are two interfaces that are core to this library, the [ResponseWriter](readers-common/src/main/java/uk/gov/gchq/palisade/reader/common/ResponseWriter.java) and the [DataReader](readers-common/src/main/java/uk/gov/gchq/palisade/reader/common/DataReader.java). 
These provide the basis of the solution for retrieving the data, filtering according to the rules for the query and provides a stream of data to be made available to the client.

The writer in the class [SerialisingResponseWriter](readers-common/src/main/java/uk/gov/gchq/palisade/reader/common/SerialisingResponseWriter.java) and the reader implementation in the abstract class [SerialisedDataReader](readers-common/src/main/java/uk/gov/gchq/palisade/reader/common/SerialisedDataReader.java) provide the implementation for the data retrieval.
The class [HadoopDataReader](hadoop-reader/src/main/java/uk/gov/gchq/palisade/reader/HadoopDataReader.java) is the reader implementation for Hadoop. In the

> [SerialisingResponseWriter.write](readers-common/src/main/java/uk/gov/gchq/palisade/reader/common/SerialisingResponseWriter.java)` method, these classes are used to

Reviewer: ```suggestion
[SerialisingResponseWriter.write](readers-common/src/main/java/uk/gov/gchq/palisade/reader/common/SerialisingResponseWriter.java) method, these classes are used to

```

In terms of code, there are two interfaces that are core to this library, the [ResponseWriter](readers-common/src/main/java/uk/gov/gchq/palisade/reader/common/ResponseWriter.java) and the [DataReader](readers-common/src/main/java/uk/gov/gchq/palisade/reader/common/DataReader.java). 
These provide the basis of the solution for retrieving the data, filtering according to the rules for the query and provides a stream of data to be made available to the client.

The writer in the class [SerialisingResponseWriter](readers-common/src/main/java/uk/gov/gchq/palisade/reader/common/SerialisingResponseWriter.java) and the reader implementation in the abstract class [SerialisedDataReader](readers-common/src/main/java/uk/gov/gchq/palisade/reader/common/SerialisedDataReader.java) provide the implementation for the data retrieval.
The class [HadoopDataReader](hadoop-reader/src/main/java/uk/gov/gchq/palisade/reader/HadoopDataReader.java) is the reader implementation for Hadoop. In the
[SerialisingResponseWriter.write](readers-common/src/main/java/uk/gov/gchq/palisade/reader/common/SerialisingResponseWriter.java) method, these classes are used to

> first retrieve the references to the data plus the rules that will be used for

Reviewer: There are a couple of returns before this line which means it's now a new paragraph within the ReadMe. I would remove at least 1 of these so that it reads properly from line 35.

Author: Changed format to the convention of one sentence per line.

# <img src="logos/logo.svg" width="180">

### Palisade Readers

The `Readers` library enables functionality for providing data with the relevant
data restrictions applied. It provides the services needed to locate the
references to the requested records, uses this to retrieve the data and then
to extract each record where it can be filtered in using the restrictions

> defined by the rules in place for the data.

Reviewer 2: Standard for Markdown *should* be semantic breaks - one-line-per-unit-of-thought - but it's a bit much.
Palisade docs in general have been using one-line-per-sentence though.

Reviewer 2: Reason being how prose and semantic writing works with git.
If i stick a big long sentence on the first line, then reformat it like this, every line thereafter changes.
You'll get more useful behaviour out of git if you write your prose like it's code - split discrete 'units' over separate lines

Author: I like being able to read the text in the editor, but yes the chopping and gluing together text is a downside.  I'll change to  the convention of a sentence per line.

The client will have sent the initial request to access data with
the Palisade Service and received a response with a unique identifier for the
request. This is then used to send a request to the Data Service to retrieve
the data. The Reader library provides the functionality for the Data Service to
retrieve the data plus the rules that need to be applied. This information is
then used to return an output stream of the filtered data to the client.

In terms of code, there are two interfaces that are core to this library, the
`ResponseWriter` and the `DataReader`. These provide the basis of the solution
for retrieving the data, filtering according to the rules for the query and
provides a stream of data to be made available to the client.

```java
public interface ResponseWriter extends AutoCloseable {
    void close() throws IOException;

    ResponseWriter write(final OutputStream output) throws IOException;
}

public interface DataReader {

    void addSerialiser(final DataFlavour flavour, final Serialiser<?> serialiser);

    DataReaderResponse read(final DataReaderRequest request, AtomicLong recordsProcessed, AtomicLong recordsReturned) throws NoCapacityException;
    
}

> ```

Reviewer 2: Reference the file rather than quote the code

Author: Replaced the code with links to the code and followed on with the logical extension of that by changing names of the files with links.

# <img src="logos/logo.svg" width="180">

### Palisade Readers

The `Readers` library enables functionality for providing data with the relevant
data restrictions applied. It provides the services needed to locate the
references to the requested records, uses this to retrieve the data and then
to extract each record where it can be filtered in using the restrictions
defined by the rules in place for the data.

A good starting point to understanding this library is with the Data Service.
The client will have sent the initial request to access data with
the Palisade Service and received a response with a unique identifier for the
request. This is then used to send a request to the Data Service to retrieve
the data. The Reader library provides the functionality for the Data Service to

> 

Reviewer: Remove this extra line break

Author: done

retrieve the data plus the rules that need to be applied. This information is
then used to return an output stream of the filtered data to the client.

> In terms of code, there are two interface that are core to this library, the

Reviewer: ```suggestion
In terms of code, there are two interfaces that are core to this library, the

```

`ResponseWriter` and the `DataReader`. These provide the basis of the solution

> 

Reviewer: Remove this extra line break

Author: done

for retrieving the data, filtering according to the rules for the query and
provides a stream of data to be made available to the client.

```java
public interface ResponseWriter extends AutoCloseable {
    void close() throws IOException;

    ResponseWriter write(final OutputStream output) throws IOException;
}

public interface DataReader {

    void addSerialiser(final DataFlavour flavour, final Serialiser<?> serialiser);

    DataReaderResponse read(final DataReaderRequest request, AtomicLong recordsProcessed, AtomicLong recordsReturned) throws NoCapacityException;
    
}

```
The writer in the class `SerialisingResponseWriter` and the reader
implementation in the abstract class `SerialisedDataReader` provide the
implementation for the data retrieval. The class `HadoopDataReader`
is the reader implementation for Hadoop. In the
`SerialisingResponseWriter#write` method, these classes are used to
first retrieve the references to the data plus the rules that will be used for
filtering. This is used to set up a processing stream for the client's request.

> The process will retrieve the data as an input stream. It will then be
> deserialised into records where the rules are applied to filter the records.

Reviewer: ```suggestion
The process will retrieve the data as an input stream. It will then be
deserialised into records where the rules are applied to redact or mask the records.
```

# <img src="logos/logo.svg" width="180">

> ### Readers

Reviewer: I feel like this should be something like "Data Readers" rather than just `Readers`

Author: How about Palisade Readers

Reviewer: This is probably the better title

The `Readers` library enables functionality for providing data with the relevant
data restrictions applied. It provides the services needed to locate the
references to the requested records, uses this to retrieve the data and then
to extract each record where it can be filtered in using the restrictions
defined by the rules in place for the data.

A good starting point to understanding this library is with the Data Service.
The client will have sent the initial request to access data with
the Palisade Service and received a response with a unique identifier for the

> request.  This is then used to send a request to the Data Service to retrieve
> data.  The Reader library provides the functionality for the Data Service to

Reviewer: ```suggestion
request. This is then used to send a request to the Data Service to retrieve
the data. The Reader library provides the functionality for the Data Service to

```

retrieve the data plus the rules that need to be applied. This information is
then used to return an output stream of the filtered data to the client.

> In terms of code, there are two interface that are core to this library:
> `ResponseWriter`; and `DataReader`. These provide the basis of the solution

Reviewer: ```suggestion
In terms of code, there are two interface that are core to this library, the
`ResponseWriter` and the `DataReader`. These provide the basis of the solution

```

for retrieving the data, filtering according to the rules for the query and
provides a stream of data to be made available to the client.

```java
public interface ResponseWriter extends AutoCloseable {
    void close() throws IOException;

    ResponseWriter write(final OutputStream output) throws IOException;
}

public interface DataReader {

    void addSerialiser(final DataFlavour flavour, final Serialiser<?> serialiser);

    DataReaderResponse read(final DataReaderRequest request, AtomicLong recordsProcessed, AtomicLong recordsReturned) throws NoCapacityException;
    
}

```

> The writer in the class `SerialisingResponseWriter` and the reader
> implementation in the abstract class `SerialisedDataReader` (plus
> it's the extended class `HadoopDataReader`) provides the Hadoop
> implementation for the data retrieval. In the

Reviewer: This reads like the writer and reader classes are geared up for implementing Hadoop. While including Hadoop is a valid idea I think it needs to be separated from the other classes.
```suggestion
The writer in the class `SerialisingResponseWriter` and the reader
implementation in the abstract class `SerialisedDataReader` provide the
implementation for the data retrieval. The class `HadoopDataReader`
is the reader implementation for Hadoop. In the
```

`SerialisingResponseWriter#write` method, these classes are used to

> first retrieve the references to the data plus the rules that will be use for
> filtering.  This is used to set up a processing stream for the client's request.

Reviewer: ```suggestion
first retrieve the references to the data plus the rules that will be used for
filtering. This is used to set up a processing stream for the client's request.

```

The process will retrieve the data as an input stream. It will then be
deserialised into records where the rules are applied to filter the records.
This is then re-serialised into an output stream.  This output stream will be
the data ready for the Data Service to provide to the client.

#### Note Configuring for Hadoop on Windows

In order to use the [HadoopResourceService](hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java) and [HadoopDataReader](hadoop-reader/src/main/java/uk/gov/gchq/palisade/reader/HadoopDataReader.java) in a Windows environment, the provided [hadoop binaries](hadoop-resource/src/test/resources/bin/) should be added to the system PATH.
If using IntelliJ's built-in test runners, these binaries will also need to be added under the project's global libraries.

> 

Reviewer: I think that this can be removed as it is covered in more detail within the services repo. Plus the link to the binaries no longer exists

Author: Some reason, I did not see this the first time I went though the comments.  Yes, this has been deleted.
